### PR TITLE
ESQL: Add boolean to track block release

### PR DESCRIPTION
**Pull request by ChrisHegarty**
This commit adds a boolean state to blocks, allowing to tell is a block has been released.

### CODE REVIEW

**Review by nik9000 — APPROVED** [review]

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java [hunk 1]
```diff
@@ -23,6 +23,8 @@ abstract class AbstractBlock implements Block {
 
     protected final BlockFactory blockFactory;
 
+    protected boolean released; // false
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:26]: `= false` isn't slower and communicates the same thing as the comment I hink.

**Review by dnhatn — APPROVED** [review]
LGTM, thanks Chris! Can we add `assert ensureNotReleased()` in Block's methods to ensure that we don't accidentally access an already-released block?

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java (hunk 1 above)
Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java]: Done

**Comment by ChrisHegarty** [comment]
> LGTM, thanks Chris! Can we add `assert ensureNotReleased()` in Block's methods to ensure that we don't accidentally access an already-released block?

Thanks for the review Nhat. It's true that adding an assert would catch potential issues early, but I'd like to try to keep the data accessor methods clean of this, so that's why we have the higher-level page check.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BasicBlockTests.java [hunk 2]
```diff
@@ -898,16 +904,36 @@ void assertZeroPositionsAndRelease(Vector vector) {
         releaseAndAssertBreaker(vector);
     }
 
-    <T extends Releasable & Accountable> void releaseAndAssertBreaker(T data) {
-        releaseAndAssertBreaker(data, breaker);
+    // <T extends Releasable & Accountable> void releaseAndAssertBreaker(T data) {
+    // releaseAndAssertBreaker(data, breaker);
+    // }
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BasicBlockTests.java:909]: Leftover?

**Review by nik9000 — APPROVED** [review]

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BasicBlockTests.java (hunk 2 above)
Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BasicBlockTests.java:909]: removed.

**Pull request merged**